### PR TITLE
[Sources] Use Immutable records

### PR DESCRIPTION
**Pull request by DavideDeCenzo**
Associated Issue: #4935

Here's the Pull Request Doc
https://devtools-html.github.io/debugger.html/CONTRIBUTING.html#pull-requests

Here's the Debugger's Testing doc
https://docs.google.com/document/d/1oBMRxV8A2ag2t22YsQOxTdEv0mXKzIg0tggJjRkU1S0/edit#.
Feel free to improve it!

### CODE REVIEW

File: src/reducers/sources.js [hunk 1]
```diff
@@ -324,7 +343,9 @@ function getSourceByUrlInSources(sources: SourcesMap, url: string) {
     return null;
   }
 
-  return sources.find(source => source.get("url") === url);
+  return sources.find(source => {
+    return source.get("url") === url;
```
[the remark this answers is not in the document]
Comment (darkwing) [reply src/reducers/sources.js]: Do you happen to know which?  Travis says everything is all good :)

Also, can we do:

sources.find(source => source.get("url") === url);

**Comment by DavideDeCenzo** [comment]
I get these errors If I put `` source.url == url `` 

testsfailed

**Comment by darkwing** [comment]
Wow, you're right; `source.get("url")` returns a URL while `source.url` return `undefined`

**Comment by DavideDeCenzo** [comment]
I've done some logs and I discovered that even the source in the function getPrettySource with `source.url` that I put returns an undefined. Actually It returns an URL with  `source.get("url").`It's strange because the tests are all passed.

sourcesundefined

sourceundefined2

mapvalue

File: src/reducers/sources.js [hunk 2]
```diff
@@ -174,8 +188,13 @@ function updateSource(state: Record<SourcesState>, source: Source | Object) {
...
     return state;
   }
+  const existingSource = state.getIn(["sources", source.id]);
 
-  return state.mergeIn(["sources", source.id], source);
+  if (existingSource) {
+    const updatedSource = existingSource.merge(source);
+    return state.setIn(["sources", source.id], updatedSource);
+  }
+  return state.mergeIn(["sources", source.id], new SourceRecordClass(source));
```
Comment (giatorta) [inline src/reducers/sources.js]: this converts the new Record into a Map, see this [immutable bug](https://github.com/facebook/immutable-js/issues/975). This causes the failures to retrieve attributes without using `get()`.

Comment (jasonLaster) [reply src/reducers/sources.js]: Thanks! is there a good work around?

**Comment by jasonLaster** [comment]
I think this PR has kinda gone out of scope. @DavideDeCenzo how do you feel about picking up something else while we try and figureout immutable.js!

File: src/reducers/sources.js (hunk 2 above)
Comment (giatorta) [reply src/reducers/sources.js]: hey @jasonLaster... I changed `state.mergeIn()` in L197 to:
`state.setIn(["sources", source.id], new SourceRecordClass(source);`
and every `source.get("x")` with `source.x` and all tests now pass! I think @DavideDeCenzo may try to finalize it

Comment (jasonLaster) [reply src/reducers/sources.js]: oh thanks. that's promising.  Re-opening

**Comment by DavideDeCenzo** [comment]
Thanks @giatorta Now It works!

**Comment by DavideDeCenzo** [comment]
@jasonLaster Now it should be better  , the tests are passed also

**Comment by jasonLaster** [comment]
We're in! Great job @DavideDeCenzo

**Pull request merged**